### PR TITLE
Adding domain

### PR DESCRIPTION
**Pull request by tianhuil**
Merges DomainScheduler implemented in https://github.com/tianhuil/domain_scheduler into scrapy. It scrapes in a domain-smart way: by round-robin cycling through the domains. This has two benefits:

1. Spreading out load on the target servers instead of hitting the server with many requests at once
2. Reducing delays caused by server-throttling or scrapy's own `CONCURRENT_REQUESTS_PER_IP` restrictions.  Empirical testing has shown this to be quite effective.

It implements the solution proposed in https://github.com/scrapy/scrapy/issues/1802#issue-135260562 which found similar performance improvements.  **Original proposal** was first posted https://github.com/scrapy/scrapy/issues/1802 (code no longer available), https://github.com/scrapy/scrapy/issues/2474, and https://github.com/scrapy/scrapy/issues/3140.  

**Note:** It requires more than just using `SCHEDULER_PRIORITY_QUEUE` as it needs an API change to the queue (passing in a non-integer key, i.e. the domain, to a new round robin queue).

- Implements https://github.com/scrapy/scrapy/issues/3140
- Depends on merged queuelib PR https://github.com/scrapy/queuelib/pull/21

Would love to get feedback @pablohoffman, @dangra, and @kmike!

### CODE REVIEW

**Comment by tianhuil** [comment]
Thanks for deploying `queuelib` to pypi @dangra.  It's now passing tests.

**Comment by kmike** [comment]
Thanks @tianhuil for pushing it forward!

But I'd like to give it more thought before merging; round-robin is an improvement over what we have now, but it is not the best solution (see e.g. https://github.com/scrapy/scrapy/issues/2474#issuecomment-288222104). Round-robin assumes requests from all domains are handled with roughly the same speed, which is not the case in a real world, esp. if you're not doing broad crawl, but crawling a specific list of domains.

File: scrapy/core/scheduler.py [hunk 1]
```diff
@@ -132,3 +141,27 @@ def _dqdir(self, jobdir):
...
+    """
+
+    def __init__(self, dupefilter, jobdir=None, dqclass=None, mqclass=None,
+                 logunser=False, stats=None, pqclass=None):
+        super(DomainScheduler, self).__init__(dupefilter, jobdir=jobdir, dqclass=dqclass,
+                                              mqclass=mqclass, logunser=logunser,
+                                              stats=stats, pqclass=RoundRobinQueue)
+
+    def request_key(cls, request):
+        return urlparse(request.url).netloc
```
Comment (kmike) [inline scrapy/core/scheduler.py]: If we go with round robin, it should be probably based on `request.meta['download_slot']`, or use the same algorithm. 

I also think that's good to have it overridable via request.meta, e.g. in https://github.com/TeamHG-Memex/linkdepth/blob/master/queues.py#L21 I used `request.meta['scheduler_slot']`. This allows to tweak these slots from spiders, and also allows to enable round-robin queue by default: scheduler_slot is empty, so all requests go to the same queue, keeping existing behavior; users who need round-robin can set a scheduler slot per-request.

File: scrapy/core/scheduler.py [hunk 2]
```diff
@@ -6,11 +6,15 @@
 from scrapy.utils.reqser import request_to_dict, request_from_dict
 from scrapy.utils.misc import load_object
 from scrapy.utils.job import job_dir
+from queuelib import RoundRobinQueue
 
 logger = logging.getLogger(__name__)
 
+def _make_file_safe(string):
+    return "".join([c if c.isalnum() or c in {'-._'} else '_' for c in string])
```
Comment (kmike) [inline scrapy/core/scheduler.py]: it seems this function is lossy - won't it make some of the requests take the same slot why the shouldn't?

**Comment by whalebot-helmsman** [comment]
@tianhuil Do you have will to work on proposed changes?

**Comment by tianhuil** [comment]
I'm happy to make the smaller changes proposed by @kmike but I thought this was held up by considerations of https://github.com/scrapy/scrapy/pull/3160#issuecomment-372788831

Let me know if I should proceed with everything except for https://github.com/scrapy/scrapy/pull/3160#issuecomment-372788831

**Comment by whalebot-helmsman** [comment]
Considerations in https://github.com/scrapy/scrapy/pull/3160#issuecomment-372788831are reasonable. At the same time. this PR is a prerequisite to address them. After this PR landed in trunk we can continue work on smarter way of scheduling requests.

**Comment by tianhuil** [comment]
OK -- I"ll respond to everything except for https://github.com/scrapy/scrapy/pull/3160#issuecomment-372788831 to help move things forward.  Happy to do it.  Give me a little bit.

File: scrapy/core/scheduler.py (hunk 2 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Agree.  But `c.isalnum() or c in {'-._'}` handles the vast majority of domain characters and any collisions will only degrade performance.  I added a comment.

File: scrapy/core/scheduler.py (hunk 1 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Thanks!  Just implemented.

**Comment by tianhuil** [comment]
Sorry for the delay.  Reponded to both of @kmike's helpful comments.

**Comment by akirmse** [comment]
Is there any news on this?  I've been eagerly awaiting this feature because I'm also seeing scrapy get stuck on a single domain in a targeted crawl.

File: scrapy/core/scheduler.py [hunk 3]
```diff
@@ -61,6 +82,7 @@ def enqueue_request(self, request):
             self._mqpush(request)
             self.stats.inc_value('scheduler/enqueued/memory', spider=self.spider)
         self.stats.inc_value('scheduler/enqueued', spider=self.spider)
+        self.stats.inc_value('scheduler/enqueued/key/{}'.format(self.request_key(request)))
```
Comment (kmike) [inline scrapy/core/scheduler.py]: Hm, I like an extra debugging options in stats, but we may have a problem if a spider is to download e.g. pages from 100K domains, one per domain - stats would become huge and unreadable.

File: scrapy/core/scheduler.py [hunk 4]
```diff
@@ -45,9 +66,9 @@ def open(self, spider):
 
     def close(self, reason):
         if self.dqs:
-            prios = self.dqs.close()
+            state = self.dqs.close()
             with open(join(self.dqdir, 'active.json'), 'w') as f:
-                json.dump(prios, f)
+                json.dump(state, f)
```
Comment (kmike) [inline scrapy/core/scheduler.py]: 👍 to rename it to state. I wonder if we should rename the file as well (though for backwards compatibility an old name should be still supported, when reading it).

File: scrapy/core/scheduler.py [hunk 5]
```diff
@@ -3,14 +3,26 @@
...
 
 logger = logging.getLogger(__name__)
 
+def _make_file_safe(string):
+    """
+    Make key file safe.
+
+    Warning: not a general function.  Works for all numbers and the vast majority of domains.
+    Collisions in file names will only degrade (midly) performance,
+    not lead to incorrect results.
```
Comment (kmike) [inline scrapy/core/scheduler.py]: I'm still not convinced we should use a function with known limitations, while there is an easy workaround - e.g. add md5 or sha1 hash (or its part?) of the original content to the end. I'm not sure, but replacing a list comprehension with a simpler regex + an md5 hash may even make this code faster. 

We also probably want to limit string length, as file systems can have problems with that.

If we're to be strict, `isalnum` check can also be problematic, as it doesn't guarantee ascii.

Anyways, I agree that the current code works in most cases, and probably it is not a big deal to leave it as-is, but it doesn't seem too hard to make it work in all cases: be more aggressive with the human-readable part, add a hash.

File: scrapy/core/scheduler.py [hunk 6]
```diff
@@ -34,6 +46,15 @@ def from_crawler(cls, crawler):
         return cls(dupefilter, jobdir=job_dir(settings), logunser=logunser,
                    stats=crawler.stats, pqclass=pqclass, dqclass=dqclass, mqclass=mqclass)
 
+    def request_key(self, request):
+        if 'scheduler_slot' in request.meta:
+            return request.meta['scheduler_slot']
```
Comment (kmike) [inline scrapy/core/scheduler.py]: Sorry for being misleading, but now I'm not sure we need `scheduler_slot`, probably reusing `download_slot` is better. Though I don't recall why exactly :) //cc @dangra do you recall our discussion?

File: scrapy/core/scheduler.py [hunk 7]
```diff
@@ -34,6 +46,15 @@ def from_crawler(cls, crawler):
...
+
+        return self._request_key(request)
+
+    def _request_key(self, request):
+        raise NotImplementedError
```
Comment (kmike) [inline scrapy/core/scheduler.py]: If this method is required, it is better to use abstract classes (see https://docs.python.org/3/library/abc.html). 

We use twisted interfaces in a few other places, which solve a similar issue, but I'd prefer to move to abc everywhere (though I don't know what others think :)

File: scrapy/core/scheduler.py (hunk 5 above)
Comment (kmike) [reply scrapy/core/scheduler.py]: Also, a nitpick: one more blank line is needed before this function according to pep8.

File: scrapy/core/scheduler.py (hunk 6 above)
Comment (Desprado666) [reply scrapy/core/scheduler.py]: 感觉你们好牛逼呀

File: scrapy/core/scheduler.py [hunk 8]
```diff
@@ -132,3 +155,27 @@ def _dqdir(self, jobdir):
             if not exists(dqdir):
                 os.makedirs(dqdir)
             return dqdir
+
+
+class DefaultScheduler(BaseScheduler):
```
[the remark this answers is not in the document]
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Done

File: docs/topics/settings.rst [hunk 9]
```diff
@@ -1106,12 +1106,14 @@ If enabled, Scrapy will respect robots.txt policies. For more information see
...
 ---------
 
-Default: ``'scrapy.core.scheduler.Scheduler'``
+Default: ``'scrapy.core.scheduler.DefaultScheduler'``
 
 The scheduler to use for crawling.
 
 .. setting:: SCHEDULER_DEBUG
 
+Can also be set to a domain-aware that round robin cycles through the domains as it outputs requests to reduce the load on servers and accelerate scraping: ``'scrapy.core.scheduler.DomainScheduler'``.
```
[the remark this answers is not in the document]
Comment (tianhuil) [reply docs/topics/settings.rst]: Done

File: scrapy/core/scheduler.py [hunk 10]
```diff
@@ -132,3 +155,27 @@ def _dqdir(self, jobdir):
...
+    """
+    Key is the priority of the request (an integer)
+    """
+
+    def _request_key(cls, request):
+        return -request.priority
+
+
+class DomainScheduler(BaseScheduler):
```
[the remark this answers is not in the document]
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Done

File: scrapy/core/scheduler.py [hunk 11]
```diff
@@ -132,3 +155,27 @@ def _dqdir(self, jobdir):
...
+    """
+    Key is the domain and we round robin.  The pqclass parameter is ignored.
```
[the remark this answers is not in the document]
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Done, except I decided not to ask the user to specify `SCHEDULER_PRIORITY_QUEUE`.

File: scrapy/core/scheduler.py [hunk 12]
```diff
@@ -3,14 +3,26 @@
 import logging
 from os.path import join, exists
 
-from scrapy.utils.reqser import request_to_dict, request_from_dict
-from scrapy.utils.misc import load_object
+from scrapy.utils.httpobj import urlparse_cached
 from scrapy.utils.job import job_dir
+from scrapy.utils.misc import load_object
+from scrapy.utils.reqser import request_to_dict, request_from_dict
+from queuelib import RoundRobinQueue
```
[the remark this answers is not in the document]
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Done.

File: scrapy/core/scheduler.py (hunk 4 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Thanks.

File: scrapy/core/scheduler.py (hunk 3 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Removed.

File: scrapy/core/scheduler.py [hunk 13]
```diff
@@ -132,3 +155,27 @@ def _dqdir(self, jobdir):
...
+    """
+
+    def __init__(self, dupefilter, jobdir=None, dqclass=None, mqclass=None,
+                 logunser=False, stats=None, pqclass=None):
+        super(DomainScheduler, self).__init__(dupefilter, jobdir=jobdir, dqclass=dqclass,
+                                              mqclass=mqclass, logunser=logunser,
+                                              stats=stats, pqclass=RoundRobinQueue)
+
+    def _request_key(cls, request):
+        return urlparse_cached(request).hostname or ''
```
[the remark this answers is not in the document]
Comment (tianhuil) [reply scrapy/core/scheduler.py]: I don't really know what to do here, perhaps because I don't know what `download_slot` is meant to do.  Perhaps you could contribute the commit here?

**Comment by tianhuil** [comment]
@kmike: let me know what your thoughts are here.

**Comment by tianhuil** [comment]
@kmike: let me know what your thoughts are: would love to get this merged.

File: docs/topics/settings.rst [hunk 14]
```diff
@@ -1106,12 +1106,15 @@ If enabled, Scrapy will respect robots.txt policies. For more information see
 SCHEDULER
 ---------
 
-Default: ``'scrapy.core.scheduler.Scheduler'``
+Default: ``'scrapy.core.scheduler.DefaultScheduler'``
```
Comment (kmike) [inline docs/topics/settings.rst]: Shouldn't it be Scheduler?

File: docs/topics/settings.rst [hunk 15]
```diff
@@ -1106,12 +1106,15 @@ If enabled, Scrapy will respect robots.txt policies. For more information see
...
 
-The scheduler to use for crawling.
+The scheduler to use for crawling.  Can also be set to a domain-aware that round
+robin cycles through the domains as it outputs requests to reduce the load on
+servers and accelerate scraping: ``'scrapy.core.scheduler.DomainScheduler'``.
```
Comment (kmike) [inline docs/topics/settings.rst]: DomainScheduler is renamed

File: scrapy/core/scheduler.py [hunk 16]
```diff
@@ -1,16 +1,31 @@
-import os
+from abc import ABC
```
Comment (kmike) [inline scrapy/core/scheduler.py]: I don't think it works in Python 2

File: scrapy/core/scheduler.py [hunk 17]
```diff
@@ -1,16 +1,31 @@
...
+
```
Comment (kmike) [inline scrapy/core/scheduler.py]: there shouldn't be a blank line between stdlib imports

File: scrapy/core/scheduler.py [hunk 18]
```diff
@@ -1,16 +1,31 @@
...
 logger = logging.getLogger(__name__)
 
 
-class Scheduler(object):
+def _make_file_safe(string):
+    """
+    Make string file safe but readable.
+    """
+    clean_string = "".join([c if c.isalnum() or c in '-._' else '_' for c in string])
+    hash_string = hashlib.md5(string).hexdigest()
```
Comment (kmike) [inline scrapy/core/scheduler.py]: this doesn't work if a string is unicode, while the code above assumes it is unicode (in Python 3)

File: scrapy/core/scheduler.py (hunk 7 above)
Comment (kmike) [reply scrapy/core/scheduler.py]: The idea was to mark this method as abstractmethod, so that it fails earlier

File: scrapy/core/scheduler.py [hunk 19]
```diff
@@ -73,6 +97,7 @@ def next_request(self):
                 self.stats.inc_value('scheduler/dequeued/disk', spider=self.spider)
         if request:
             self.stats.inc_value('scheduler/dequeued', spider=self.spider)
+            self.stats.inc_value('scheduler/dequeued/key/{}'.format(self.request_key(request)))
```
Comment (kmike) [inline scrapy/core/scheduler.py]: if `enqueued` stat is removed, this one should be removed as well

File: scrapy/core/scheduler.py [hunk 20]
```diff
@@ -99,28 +124,28 @@ def _dqpush(self, request):
...
                 return request_from_dict(d, self.spider)
 
-    def _newmq(self, priority):
+    def _newmq(self, key):
         return self.mqclass()
 
-    def _newdq(self, priority):
-        return self.dqclass(join(self.dqdir, 'p%s' % priority))
+    def _newdq(self, key):
+        return self.dqclass(join(self.dqdir, _make_file_safe('k%s' % key)))
```
Comment (kmike) [inline scrapy/core/scheduler.py]: hm, actually this is backwards incompatible - after this PR scrapy won't be able to resume crawls from earlier Scrapy versions

File: scrapy/core/scheduler.py [hunk 21]
```diff
@@ -132,3 +157,33 @@ def _dqdir(self, jobdir):
...
+    the domain that was requested last.  The default Scheduler sends multiple
+    consecutive requests to a domain if multiple links were discovered and
+    added consecutively.  This scheduler spreads those out that workload.
+
+    Uses `RoundRobinQueue`.  The pqclass parameter and SCHEDULER_PRIORITY_QUEUE
+    are ignored.
+    """
+
+    def __init__(self, dupefilter, jobdir=None, dqclass=None, mqclass=None,
+                 logunser=False, stats=None, pqclass=None):
```
Comment (kmike) [inline scrapy/core/scheduler.py]: I think it should be possible to override `pqclass` argument somehow, now it is silently discarded

File: scrapy/settings/default_settings.py [hunk 22]
```diff
@@ -241,7 +241,7 @@
 
 ROBOTSTXT_OBEY = False
 
-SCHEDULER = 'scrapy.core.scheduler.Scheduler'
+SCHEDULER = 'scrapy.core.scheduler.DefaultScheduler'
```
Comment (kmike) [inline scrapy/settings/default_settings.py]: there is no DefaultScheduler class

**Comment by kmike** [comment]
hey @tianhuil, I've left some more comments. Besides them, there are the following issues:

1. Merge conflicts should be resolved.
2. Tests should pass.
3. It is still not documented that request priorities don't work. I'd prefer them working, but if they don't, it should be clearly documented, both in scheduler docs and in Request docs.

File: scrapy/core/scheduler.py (hunk 18 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: fixed

File: scrapy/core/scheduler.py (hunk 19 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Good point.

File: scrapy/core/scheduler.py (hunk 20 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: I put this into the base class.

File: scrapy/core/scheduler.py (hunk 21 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: That's a good point.  I just removed it from the signature of the child class.

File: scrapy/settings/default_settings.py (hunk 22 above)
Comment (tianhuil) [reply scrapy/settings/default_settings.py]: Fixed

File: scrapy/core/scheduler.py (hunk 16 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Don't know enough about `ABC`.  Perhaps we can just leave out this part and have it inherit from `object`?

**Comment by tianhuil** [comment]
@kmike: it's I've documented that priorities have not been used.  Looks like it's passing.

File: scrapy/core/scheduler.py (hunk 7 above)
Comment (tianhuil) [reply scrapy/core/scheduler.py]: Fixed.

**Comment by tianhuil** [comment]
@kmike: let me know if you want to merge this.  Would like to do it before conflicts re-emerge.

**Comment by tianhuil** [comment]
@kmike: let me know if you have any other questions or comments. Would like to do it before conflicts re-emerge.

**Comment by tianhuil** [comment]
@kmike: let me know if you have any other questions or comments. Would like to do it before conflicts re-emerge.

File: scrapy/core/scheduler.py [hunk 23]
```diff
@@ -1,16 +1,31 @@
...
+    """
+    Make string file safe but readable.
+    """
+    clean_string = "".join([c if c.isalnum() or c in '-._' else '_' for c in string])
+    hash_string = hashlib.md5(string.encode('utf8')).hexdigest()
+    return "{}-{}".format(clean_string[:40], hash_string[:10])
+
+
+class BaseScheduler(object):
+    __metaclass__ = abc.ABCMeta
```
Comment (kmike) [inline scrapy/core/scheduler.py]: this doesn't work in Python 3. Please use [six.with_metaclass](https://pythonhosted.org/six/#six.with_metaclass)

**Comment by kmike** [comment]
Hey @tianhuil! Sorry for going back and forth. 

My main concern is a lack of support for request priorities. There are implementations which support request priorities - in fact, all other implementations I'm aware of (mentioned in previous tickets) support request priorities - they are incomplete in different ways though. 

It is not only a user-facing feature, to allow users to set custom priorities for different Requests (`scrapy.Request(url, self.parse, priority=5`). Scrapy uses request priorities itself, internally - they are used for retries and redirects; they are also a base of DepthMiddleware. 

Not having priority support may cause strange issues: for example, redirects won't necessarily be on top of the queue, so any redirect can potentially go to the end of the crawl, and, depending on a spider, may be never crawled - it seems to be the case if user uses FIFO queues, as recommended in FAQ [section](https://doc.scrapy.org/en/latest/faq.html#does-scrapy-crawl-in-breadth-first-or-depth-first-order) on how to configure BFO crawl. Actually, BFO won't work at all, as DEPTH_PRIORITY won't work.

Documentation should be updated in several more places, maybe even regardless of the implementation:

* Request [docs](https://doc.scrapy.org/en/latest/topics/request-response.html#scrapy.http.Request), 
* [Broad Crawls](https://doc.scrapy.org/en/latest/topics/broad-crawls.html), 
* this FAQ [entry](https://doc.scrapy.org/en/latest/faq.html#does-scrapy-crawl-in-breadth-first-or-depth-first-order).

**Comment by tianhuil** [comment]
@kmike: can you provide a clear set of specifications you want this PR to meet?  I'm happy to work on this but I don't have time to chase a moving target.

**Comment by kmike** [comment]
@tianhuil I'm fine with almost everything which makes things better and keeps the existing features working. I haven't realized at first that not supporting priorities is not just a problem with `priority` argument, it is also a problem for retries, redirects, BFO, scrapy-splash, etc - sorry about that. So I think RoundRobin with support for priorities is good. There are several ways to support priorities in Round-Robin:

1. Priorities are respected for each domain, but domain is chosen in a round-robin fashion.
2. Only requests of the same priority are chosen in a round-robin fashion; priorities are respected globally.

(1) makes more sense to me as a default. We just need to make sure it is clearly documented.

---

Eventually we should be aiming for an even more efficient scheduler. From my point of view, we're fixing a bug here: Scrapy may under-utilize the downloader by filling it with requests it can't download because of its own concurrency limits. 

Round-robin is a solution, but a partial one: it assumes all domains can be downloaded with the same speed. It is not always the case. For example, let's say we have domains A and B, both with many requests in a queue, and request can be finished twice as fast for domain A. Global concurrency is set to 4, and per-domain concurrency is limited to 2. With Round-Robin downloader queue will be filled like this:

1. `____ -> ABAB`: requests are added in a round-robin fashion, total size is 4.
2. `_BAB -> ABAB`: A is downloaded, next request is for A, we add it to the queue
3. `AB_B -> ABBB`: A is downloaded (it is twice as fast), B is added according to round-robin rules
4. `A_BB -> AABB`: B is downloaded, we add A
5. `_ABB -> BABB`: A is downloaded, we add B
6. `B_BB -> BABB`: A is downloaded, we add A
7. `BAB_ -> BABB`: B is downloaded, we add B
...

Note how we almost always have less A requests in the downloader queue - one instead of optimal two; it means in this case there are no parallel requests to domain A, while user requested 2 parallel requests. At the same time, there are more B requests than can be downloaded in parallel (as the limit is 2), they're just sitting in the queue, taking slots which could have been used for other domains.

In broad crawls it is rather common to get some domains which can timeout on all requests. Maybe some DNS issues. Default Scrapy timeout is 180s; with 2 retries it is 180*3=540s to process a request in a worst case. If for the "average" domain speed is 1s, and we download 100 domains in parallel, we may easily get downloader queue filled 4/5 (or so, I haven't estimated it properly) with requests from this slow domain, leaving less opportunity for parallelization of other domains, and so crawling slower than possible for no good reason.

To solve this issue scheduler should avoid adding request to the downloader if there is no free slot for it. This is how it works in frontera. Round Robin can be modified to do the same - and that's the reason I'm fine with merging round-robin implementation - it can be modified to support this feature in future.  For example, it can keep track of the current downloader state using `request_reached_downloader` (see https://github.com/scrapy/scrapy/pull/3393) and other signals, and skip adding a request from a domain if adding it would make Downloader go over concurrency limit for this domain. Instead of signals we may also query Downloader status directly.

**Comment by akirmse** [comment]
Hello, is there any progress on this?  We have an application where we have a targeted crawl over a small number of domains.  Performance is currently very poor, as the queue quickly gets filled with a single slow domain.

We don't need priorities or anything fancy.  A limited, 90% solution today would be infinitely preferable to a perfect 100% solution in a year (this review has already taken 7 months).

**Comment by kmike** [comment]
hey @akirmse! There are third-party solutions which you can use without changing Scrapy. For example, you can use RoundRobinPriorityQueue from https://github.com/TeamHG-Memex/linkdepth/blob/master/queues.py:

```
SCHEDULER_PRIORITY_QUEUE = 'queues.RoundRobinPriorityQueue'
SCHEDULER_DISK_QUEUE = 'queues.DiskQueue'
```

It actually supports priorities (so all Scrapy features works properly), but has other limitations - crawl resume is not supported. Priorities is not something fancy, they are needed for e.g. for retries, redirects, BFO crawling, scrapy-splash to work properly.

There are also other workarounds, e.g. using frontera. It is not like this is impossible to do with Scrapy, and it is not even a lot of code to get a partial solution which doesn't need any changes in Scrapy core. That's why I'd prefer to have something proper, integrated well with all Scrapy features, if we're going to make a change in Scrapy itself.

**Comment by amonroe** [comment]
@kmike Can this be merged in? @tianhuil makes scrapy usable for more than one domain.  The current version is unusable due to the scheduler.  As this isn't the default scheduler, declaring it beta and getting usage would solve a lot of issues that have existed for years.

**Comment by kmike** [comment]
@amonroe https://github.com/scrapy/scrapy/pull/3520 is much closer to get merged at the moment; it solves the same problem, but covers more edge cases, has more tests, and has better behavior - looking for downloader activity instead of using round-robin can speed up crawling significantly in many practical cases.

**Comment by amonroe** [comment]
Thanks @kmike.  I'll try that PR as well.  So far #3160 has started to show some (not much) balancing across domains after having been blackholed on one for a day.  Prior backlog is probably impacting results.   No issues as of yet.

**Comment by kmike** [comment]
Closing it, as https://github.com/scrapy/scrapy/pull/3520 is merged. Thanks @tianhuil for getting the ball rolling!
